libflux/src/semantic/tests.rs

```diff
@@ -230,6 +230,748 @@ fn instantiation_1() {
+            "a" => "forall [] int",
+            "b" => "forall [] int",
+        ],
+        src: r#"
+            c = a ^ b
+        "#,
+        exp: map![
+            "c" => "forall [] int",
+        ],
+    }
+    test_infer_err! {
+        env: map![
+            "a" => "forall [] float",
+            "b" => "forall [] int",
+        ],
+        src: r#"
+            c = a ^ b
+        "#,
+    }
+//    test_infer! {
+//        env: map![
+//            "a" => "forall [] t0",
+//            "b" => "forall [] int",
+//        ],
+//        src: r#"
+//            c = a ^ b
+//        "#,
+//        exp: map![
+//            "a" => "forall [] int",
+//            "c" => "forall [] int",
+//        ],
+//    }
+}
+#[test]
+fn binary_expr_modulo() {
+//    test_infer! {
+//        env: map![
+//            "a" => "forall [] t0",
+//            "b" => "forall [] t1",
+//        ],
+//        src: r#"
+//            c = a % b
+//        "#,
+//        exp: map![
+//            "a" => "forall [] t1",
+//            "b" => "forall [] t1",
+//            "c" => "forall [] t1",
+//        ],
+//    }
+    test_infer!{
+        env: map![
+            "a" => "forall [] int",
+            "b" => "forall [] int",
+        ],
+        src: r#"
+            c = a % b
+        "#,
+        exp: map![
+            "c" => "forall [] int",
+        ],
+    }
+    test_infer_err! {
+        env: map![
+            "a" => "forall [] float",
+            "b" => "forall [] int",
+        ],
+        src: r#"
+            c = a % b
+        "#,
+    }
+//    test_infer! {
+//        env: map![
+//            "a" => "forall [] t0",
+//            "b" => "forall [] int",
+//        ],
+//        src: r#"
+//            c = a % b
+//        "#,
+//        exp: map![
+//            "a" => "forall [] int",
+//            "b" => "forall [] int",
+//            "c" => "forall [] int",
+//        ],
+//    }
+}
+#[test]
+fn binary_expr_comparison() {
+    for op in vec![">=", "<=", ">", "<", "==", "!="] {
+        let src = format!("c = a {} b", op);
+//        test_infer! {
+//            env: map![
+//                "a" => "forall [] t0",
+//                "b" => "forall [] t1",
+//            ],
+//            src: &src,
+//            exp: map![
+//                "c" => "forall [] bool",
+//            ],
+//        }
+        test_infer! {
+            env: map![
+                "a" => "forall [] int",
+                "b" => "forall [] int",
+            ],
+            src: &src,
+            exp: map![
+                "c" => "forall [] bool",
+            ],
+        }
+        test_infer! {
+            env: map![
+                "a" => "forall [] float",
+                "b" => "forall [] int",
+            ],
+            src: &src,
+            exp: map![
+                // TODO(jsternberg): This should be an error
+                // when polymorphic literals are implemented.
+                "c" => "forall [] bool",
+            ],
+        }
```

Comment:
This would pass if we had polymorphic literals. It should fail now though.

Reply:
This needs to be tested. It's in the Go code as something that is allowed.

Reply:
Also, this should not pass with polymorphic literals. These are explicit float and integer types. This isn't an example of two literals being compared.

Reply:
This is correct, this specific test case would not pass with polymorphic literals. What I meant to say was that you may compare an integer literal and a float literal with polymorphic literals. However without them, the type system should not allow it.

I realize this is currently in the Go code, but keep in mind there are many inconsistencies in the current type system code. This is one of the areas where the Go code is not correct. This type system is considered a breaking change for this very reason and is why we need to define how flux versioning works before promoting it. So it is entirely safe and reasonable to implement these constraints correctly. We do not have to maintain parity with the Go code.

Reply:
Ok. I'll change it. But I think that means we're going to need to implement polymorphic literals before this gets promoted to be used as the primary type system. I don't think we can promote a type system that doesn't allow the behavior that's meant to be allowed here. It would be safer to allow this and then constrain it later once polymorphic literals exist because then there are fewer prerequisites to integrating this into the Go code.